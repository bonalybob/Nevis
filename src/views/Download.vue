<template>
  <base-layout>
    <div slot="menu">
      <label>Port:</label>
      <button :class="{ dropdown: true, nohover: connected }" @click="refreshPortList()">
        <p>{{ selectedPort }}</p>
        <svg v-if="!connected" viewBox="0 0 24 24">
          <path d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"></path>
          <path d="M0-.75h24v24H0z" fill="none"></path>
        </svg>
      </button>
      <ul v-if="portOpen">
        <li v-for="port of portList" :key="port" @click="changePort(port)">{{ port }}</li>
      </ul>
      <label>Baud:</label>
      <button :class="{ dropdown: true, nohover: connected }" @click="refreshBaudList()">
        <p>{{ selectedBaud }}</p>
        <svg v-if="!connected" viewBox="0 0 24 24">
          <path d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"></path>
          <path d="M0-.75h24v24H0z" fill="none"></path>
        </svg>
      </button>
      <ul v-show="baudOpen">
        <li v-for="baud of baudList" :key="baud" @click="changeBaud(baud)">{{ baud }}</li>
      </ul>
      <button @click="connect()">{{ connectButtonText }}</button>
      <button @click="back" class="back">Back</button>
    </div>
    <div slot="main" class="main">
      <div v-show="lastDownload" class="card">
        <h1>Last Download</h1>
        <p v-if="lastDownload.name">Name: {{ lastDownload.name }}</p>
        <p>SI Card: {{ lastDownload.siid }}</p>
        <p>Start: {{ $time.actual(lastDownload.start) || '-'}}</p>
        <p>Finish: {{ $time.actual(lastDownload.finish) || '-'}}</p>
        <p>Time: {{ $time.elapsed(time)}}</p>
      </div>
    </div>
  </base-layout>
</template>

<script>
import BaseLayout from '@/components/BaseLayout'
import DropdownInput from '@/components/DropdownInput'

export default {
  components: {
    'base-layout': BaseLayout,
    'dropdown-input': DropdownInput,
  },

  data: () => ({
    selectedPort: '',
    selectedBaud: 38400,
    portList: [],
    baudList: [4800, 38400],
    portOpen: false,
    baudOpen: false,
    connected: false,
    connectButtonText: 'Connect',
    lastDownload: false,
  }),

  created: function () {
    if (this.$database.database === null) {
      this.$router.push('/')
      this.$messages.addMessage('Not Connected to the Database', 'error')
    }
  },

  methods: {
    back: function () {
      this.$port.disconnect()
      this.$router.go(-1)
    },
    refreshPortList: function () {
      if (!this.connected) {
        this.$port.listPorts().then(ports => {
          if (typeof ports === 'string') this.$messages.addMessage(ports, 'error')
          else this.portList = ports
          this.portOpen = !this.portOpen
        })
      }
    },
    refreshBaudList: function () {
      if (!this.connected) this.baudOpen = !this.baudOpen
    },
    changeBaud: function (baud) {
      this.selectedBaud = baud
      this.baudOpen = !this.baudOpen
    },
    changePort: function (port) {
      if (port === 'No Ports Found') this.selectedPort = ''
      else this.selectedPort = port
      this.portOpen = !this.portOpen
    },
    connect: function () {
      if (this.selectedPort === '') {
        this.$messages.addMessage('Please Select a Port to Connect to', 'error')
      }
      else if (this.connectButtonText === 'Disconnect') {
        this.$port.disconnect()
        this.connectButtonText = 'Connect'
      }
      else {
        this.$port.open = () => {
          this.connectButtonText = 'Disconnect'
          this.connected = true
        }
        this.$port.close = () => {
          this.connectButtonText = 'Connect'
          this.connected = false
        }
        this.$port.data = data => {
          this.$si.getInfo(data, this.$port.port)
            .then(data => {
              if (data) this.saveCardData(data)
            })
            .catch(error => this.$messages.addMessage(error.message, 'error'))
        }
        this.$port.error = error => this.$messages.addMessage(error, 'error')

        this.$port.connect(this.selectedPort, this.selectedBaud)
      }
    },
    saveCardData: function (data) {
      this.lastDownload = data
      this.$database.findCompetitorBySIID(data.siid)
        .then(competitor => {
          if (competitor) {
            this.lastDownload.name = competitor.name
            competitor.download = data
            this.$database.updateCompetitor(competitor)
          }
          else {
            const competitor = {
              name: data.name || 'Unknown',
              siid: data.siid.toString(),
              ageClass: '',
              club: '',
              course: 'Unknown',
              membershipNumber: '',
              nonCompetitive: false,
              download: data,
            }
            this.$database.addCompetitor(competitor)
          }
        })
    },
  },

  computed: {
    time: function () {
      if (this.lastDownload) return this.lastDownload.finish - this.lastDownload.start
    }
  },
}
</script>
<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'

label
  display: block
  margin: 3px 0
  width: 100%
  color: main-color
  text-align: center
  font-size: 14px
  default-font()

  &:first-child:not()
    margin-top: 8px

button
  margin-top: 8px

  &.dropdown
    display: flex
    justify-content: space-between
    margin-top: 0
    padding: 0
    height: 33px
    vertical-align: middle
    text-align: center

    svg
      margin: 0 4.5px
      width: 24px
      height: 33px
      vertical-align: middle
      transition: 0.3s ease-out
      fill: main-color

    p
      margin: auto
      width: calc(100% - 33px)
      vertical-align: middle
      text-align: center

    &:hover
      background-color: main-color

      svg
        fill: white

ul
  margin: 0
  padding: 0
  list-style: none

  li
    padding: 5px 0
    height: 20px
    color: main-color
    vertical-align: middle
    text-align: center
    transition: 0.3s ease-out
    default-font()

    &:hover
      background-color: main-color
      color: white

.nohover
  background-color: white
  color: main-color

  &:hover
    background-color: white !important
    color: main-color !important
</style>
