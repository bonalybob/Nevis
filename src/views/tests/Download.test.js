import { shallowMount } from '@vue/test-utils'
import Download from '@/views/Download'
import si from '@/scripts/si/si'
import courseMatching from '@/scripts/courseMatching/courseMatching'

jest.mock('@/scripts/si/si')
jest.mock('@/scripts/courseMatching/courseMatching')

test('Is a Vue Instance', () => {
  global.window.require = jest.fn()
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  global.window.require = jest.fn()
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  wrapper.setData({
    lastDownload: {
      name: 'Bob',
      courseName: 'Long',
      time: 'M3',
      siid: '12345',
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: false,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Exit', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  wrapper.vm.exit()
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12')
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  wrapper.setData({ connected: true })
  wrapper.vm.exit()
  expect(wrapper.vm.showConfirmationDialog).toBeTruthy()
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
})

test('Leave Page', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  wrapper.setData({ port: { close: jest.fn() } })
  wrapper.vm.leavePage(true)
  expect(wrapper.vm.port.close).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12')
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  wrapper.vm.leavePage(false)
  expect(wrapper.vm.port.close).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12')
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
})

test('Selected Baud', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  wrapper.setData({ selectedBaudString: 'USB (38400)' })
  expect(wrapper.vm.selectedBaud).toBe(38400)
  wrapper.setData({ selectedBaudString: 'Serial (4800)' })
  expect(wrapper.vm.selectedBaud).toBe(4800)
})

test('Port Functions', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  wrapper.vm.portOnOpen()
  expect(wrapper.vm.connected).toBeTruthy()
  expect(wrapper.vm.$messages.clearMessages).toHaveBeenCalledTimes(1)
  wrapper.vm.portOnClose()
  expect(wrapper.vm.connected).toBeFalsy()
  wrapper.vm.portOnError({ message: 'error' })
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'error',
    'error'
  )
  wrapper.setMethods({ saveDownload: jest.fn() })

  si.parseData = jest.fn(() => 'a')
  wrapper.vm.portOnData()
  expect(wrapper.vm.saveDownload).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.saveDownload).toHaveBeenLastCalledWith('a')
  si.parseData = jest.fn(() => false)
  wrapper.vm.portOnData()
  expect(wrapper.vm.saveDownload).toHaveBeenCalledTimes(1)

  si.parseData = jest.fn(() => {
    throw Error()
  })
  wrapper.vm.portOnData()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Saving Download',
    'error'
  )
})

test('Refresh Ports List - Success', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest
          .fn()
          .mockResolvedValue([{ comName: 'COM1' }, { comName: 'COM2' }]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  await wrapper.vm.refreshPortsList()
  expect(wrapper.vm.portsList).toEqual(['COM1', 'COM2'])
})

test('Refresh Ports List - No Ports', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  await wrapper.vm.refreshPortsList()
  expect(wrapper.vm.portsList).toEqual(['No Ports Found'])
})

test('Refresh Ports List - Error', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  await wrapper.vm.refreshPortsList()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Finding Ports',
    'error'
  )
})

test('Get Course From Id', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest
          .fn()
          .mockResolvedValue([{ name: 'Hello', controls: '1,2,,3' }]),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  expect(await wrapper.vm.getCourseFromId(1)).toEqual({
    name: 'Hello',
    controls: ['1', '2', '3'],
  })
})

test('Get Course From Id - Error', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  await wrapper.vm.getCourseFromId(1)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Courses',
    'error'
  )
})

test('Get Courses', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue([
          { name: 'Hello', controls: '1,2,,3' },
          { name: 'Alpha', controls: '101,525,232' },
        ]),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  expect(await wrapper.vm.getCourses()).toEqual([
    { name: 'Hello', controls: ['1', '2', '3'] },
    { name: 'Alpha', controls: ['101', '525', '232'] },
  ])
})

test('Get Courses - Error', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Courses',
    'error'
  )
})

test('Calculate Time', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })

  expect(
    wrapper.vm.calculateTime({ errors: '' }, [
      { controlCode: 'S', time: 10 },
      { controlCode: 'F', time: 19 },
    ]).displayTime
  ).toBe('00:09')

  expect(
    wrapper.vm.calculateTime({ errors: '' }, [
      { controlCode: 'S', time: 10 },
      { controlCode: 'F', time: 130 },
    ]).displayTime
  ).toBe('02:00')

  expect(
    wrapper.vm.calculateTime({ errors: '' }, [
      { controlCode: 'S', time: 10 },
      { controlCode: 'F', time: 139 },
    ]).displayTime
  ).toBe('02:09')

  expect(
    wrapper.vm.calculateTime({ errors: 'M1' }, [
      { controlCode: 'S', time: 10 },
      { controlCode: 'F', time: 19 },
    ]).displayTime
  ).toBe('M1')

  expect(
    wrapper.vm.calculateTime({ errors: '' }, [{ controlCode: 'F', time: 19 }])
      .displayTime
  ).toBe('MS')

  expect(
    wrapper.vm.calculateTime({ errors: 'M1' }, [{ controlCode: 'F', time: 19 }])
      .displayTime
  ).toBe('MS M1')

  expect(
    wrapper.vm.calculateTime({ errors: '' }, [{ controlCode: 'S', time: 10 }])
      .displayTime
  ).toBe('Rtd')

  expect(
    wrapper.vm.calculateTime({ errors: 'M1' }, [{ controlCode: 'S', time: 10 }])
      .displayTime
  ).toBe('Rtd')
})

test('Connect to Port - Already Open', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  wrapper.setData({ connected: true, port: { close: jest.fn() } })
  wrapper.vm.connect()
  expect(wrapper.vm.port.close).toHaveBeenCalledTimes(1)
})

test('Connect to Port - No Port Selected', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
        on: jest.fn(),
      },
    },
  })
  wrapper.setData({ connected: false, selectedPort: '' })
  wrapper.vm.connect()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'No Port Selected',
    'error'
  )
})

test('Connect to Port - Open', () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: function() {
        this.list = jest.fn().mockRejectedValue()
        this.open = jest.fn()
        this.close = jest.fn()
        this.on = jest.fn()
      },
    },
    methods: {
      refreshPortsList: jest.fn(),
    },
  })
  wrapper.setData({ connected: false, selectedPort: 'COM1' })
  wrapper.vm.connect()
  expect(wrapper.vm.port.on).toHaveBeenCalledTimes(4)
})

test('Find Competitor For Download', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn() },
      $archive: { connection: {}, connected: true, query: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
        on: jest.fn(),
      },
    },
  })

  wrapper.vm.$database.query.mockResolvedValue([
    { siid: '123', downloaded: false, name: 'person' },
  ])
  wrapper.vm.$archive.query.mockResolvedValue([])
  expect(await wrapper.vm.findCompetitorForDownload({ siid: '123' })).toEqual({
    siid: '123',
    downloaded: false,
    name: 'person',
  })

  wrapper.vm.$database.query
    .mockResolvedValueOnce([
      { id: 6, siid: '123', downloaded: true, name: 'person' },
    ])
    .mockResolvedValueOnce([{ controlCode: 'S', time: 523 }])
  wrapper.vm.$archive.query.mockResolvedValue([])
  expect(
    await wrapper.vm.findCompetitorForDownload({
      siid: '123',
      punches: [
        { controlCode: 'S', time: 523 },
        { controlCode: '101', time: 452 },
      ],
    })
  ).toEqual({ id: 6, siid: '123', downloaded: true, name: 'person' })
  expect(wrapper.vm.$database.query).toHaveBeenLastCalledWith(
    `DELETE FROM punches WHERE competitor=?`,
    6
  )

  wrapper.vm.$database.query
    .mockResolvedValueOnce([
      { id: 6, siid: '123', downloaded: true, name: 'person' },
    ])
    .mockResolvedValueOnce([{ controlCode: 'S', time: 623 }])
  wrapper.vm.$archive.query.mockResolvedValue([])
  expect(
    await wrapper.vm.findCompetitorForDownload({
      siid: '123',
      punches: [{ controlCode: 'S', time: 523 }],
    })
  ).toEqual({ name: 'Unknown', event: 12, siid: '123', downloaded: false })
  expect(wrapper.vm.$database.query).toHaveBeenLastCalledWith(
    'INSERT INTO competitors SET ?',
    expect.any(Object)
  )

  wrapper.vm.$database.query
    .mockResolvedValueOnce([])
    .mockResolvedValueOnce({ insertId: 5 })
  wrapper.vm.$archive.query.mockResolvedValue([
    { name: 'Bob', siid: '123', gender: 'm', yearOfBirth: '1989' },
  ])
  expect(
    await wrapper.vm.findCompetitorForDownload({
      siid: '123',
      punches: [{ controlCode: 'S', time: 523 }],
    })
  ).toEqual({
    ageClass: 'M21',
    club: undefined,
    downloaded: false,
    event: 12,
    id: 5,
    membershipNumber: undefined,
    name: 'Bob',
    siid: '123',
  })
  expect(wrapper.vm.$database.query).toHaveBeenLastCalledWith(
    'INSERT INTO competitors SET ?',
    expect.any(Object)
  )

  wrapper.vm.$database.query
    .mockResolvedValueOnce([])
    .mockResolvedValueOnce({ insertId: 5 })
  wrapper.vm.$archive.query.mockResolvedValue([])
  expect(
    await wrapper.vm.findCompetitorForDownload({
      siid: '123',
      punches: [{ controlCode: 'S', time: 523 }],
      personalData: ['Bob', 'Jones', 'm', '1989', 'HAT'],
    })
  ).toEqual({
    ageClass: 'M21',
    club: 'HAT',
    downloaded: false,
    event: 12,
    id: 5,
    membershipNumber: undefined,
    name: 'Bob Jones',
    siid: '123',
  })
  expect(wrapper.vm.$database.query).toHaveBeenLastCalledWith(
    'INSERT INTO competitors SET ?',
    expect.any(Object)
  )
})

test('Save Download', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $archive: { connection: {}, connected: true, query: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
        on: jest.fn(),
      },
    },
    methods: {
      findCompetitorForDownload: jest
        .fn()
        .mockResolvedValue({ id: 6, name: 'Bob', siid: '123', course: 3 }),
      getCourseFromId: jest
        .fn()
        .mockResolvedValue({ name: 'Long', controls: [], id: 3 }),
    },
  })
  jest.spyOn(wrapper.vm, 'getCourseFromId')
  courseMatching.linear.mockReturnValue({
    percentageCorrect: 1,
    errors: '',
    links: [],
  })
  await wrapper.vm.saveDownload({
    punches: [
      { controlCode: 'S', time: 5 },
      { controlCode: '101', time: 32 },
      { controlCode: 'F', time: 35 },
    ],
  })
  expect(wrapper.vm.$database.query).toBeCalledWith(
    'INSERT INTO punches (controlCode, time, competitor, event) VALUES ?',
    [
      [
        ['S', 5, 6, 12],
        ['101', 32, 6, 12],
        ['F', 35, 6, 12],
      ],
    ]
  )
  expect(wrapper.vm.getCourseFromId).toHaveBeenCalledTimes(1)
  expect(
    wrapper.vm.$database.query
  ).toBeCalledWith('UPDATE competitors SET ? WHERE id=?', [
    { downloaded: true },
    6,
  ])
  expect(wrapper.vm.$database.query).toBeCalledWith(
    'REPLACE INTO results SET ?',
    {
      time: 30,
      links: expect.any(String),
      errors: '',
      competitor: 6,
      event: 12,
    }
  )
  expect(wrapper.vm.lastDownload).toEqual({
    id: 6,
    name: 'Bob',
    siid: '123',
    courseName: 'Long',
    time: '00:30',
    course: 3,
  })
})

test('Save Download - Unknown Course', async () => {
  const wrapper = shallowMount(Download, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $archive: { connection: {}, connected: true, query: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockRejectedValue(),
        open: jest.fn(),
        close: jest.fn(),
        on: jest.fn(),
      },
    },
    methods: {
      findCompetitorForDownload: jest
        .fn()
        .mockResolvedValue({ id: 6, name: 'Bob', siid: '123' }),
      getCourses: jest.fn(),
    },
  })

  courseMatching.linear.mockReturnValue({ percentageCorrect: 1, errors: '' })
  courseMatching.findBestCourse.mockReturnValue({
    name: 'Long',
    controls: [],
    id: 3,
  })
  await wrapper.vm.saveDownload({
    punches: [
      { controlCode: 'S', time: 5 },
      { controlCode: '101', time: 32 },
      { controlCode: 'F', time: 35 },
    ],
  })
  expect(wrapper.vm.$database.query).toBeCalledWith(
    'INSERT INTO punches (controlCode, time, competitor, event) VALUES ?',
    [
      [
        ['S', 5, 6, 12],
        ['101', 32, 6, 12],
        ['F', 35, 6, 12],
      ],
    ]
  )
  expect(
    wrapper.vm.$database.query
  ).toBeCalledWith('UPDATE competitors SET ? WHERE id=?', [{ course: 3 }, 6])
  expect(
    wrapper.vm.$database.query
  ).toBeCalledWith('UPDATE competitors SET ? WHERE id=?', [
    { downloaded: true },
    6,
  ])
  expect(wrapper.vm.lastDownload).toEqual({
    id: 6,
    name: 'Bob',
    siid: '123',
    courseName: 'Long',
    time: '00:30',
  })
})
