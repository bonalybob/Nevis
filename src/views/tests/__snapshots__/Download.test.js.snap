// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<main>
  <back-arrow-stub
    disable="true"
    to=""
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    
    Download
  
  </h1>
   
  <div
    class="mx-12 mb-3 my-shadow"
  >
    <dropdown-input-stub
      label="Port:"
      list="No Ports Found"
      value=""
    />
     
    <dropdown-input-stub
      label="Baud:"
      list="Serial (4800),USB (38400)"
      value="USB (38400)"
    />
  </div>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Connect
    </button>
  </div>
   
  <div
    class="mx-12 mb-3 my-shadow px-3 pt-2 pb-1"
  >
    <h2
      class="mb-1"
    >
      Last Download
    </h2>
     
    <p>
      <b>
        Name:
      </b>
      
      Bob
    
    </p>
     
    <p>
      <b>
        SI Card:
      </b>
      
      12345
    
    </p>
     
    <p>
      <b>
        Course:
      </b>
      
      Long
    
    </p>
     
    <p>
      <b>
        Time:
      </b>
      
      M3
    
    </p>
  </div>
   
  <!---->
</main>
`;
